File: SignalUtilitiesKit/Screen Lock/ScreenLockViewController.m
//
//  Copyright (c) 2019 Open Whisper Systems. All rights reserved.
//

#import "ScreenLockViewController.h"
#import "UIColor+OWS.h"
#import "UIFont+OWS.h"
#import "UIView+OWS.h"
#import <SignalUtilitiesKit/SignalUtilitiesKit-Swift.h>
#import <SessionUIKit/SessionUIKit.h>

NSString *NSStringForScreenLockUIState(ScreenLockUIState value)
{
    switch (value) {
        case ScreenLockUIStateNone:
            return @"ScreenLockUIStateNone";
        case ScreenLockUIStateScreenProtection:
            return @"ScreenLockUIStateScreenProtection";
        case ScreenLockUIStateScreenLock:
            return @"ScreenLockUIStateScreenLock";
    }
}

@interface ScreenLockViewController ()

@property (nonatomic) UIView *screenBlockingImageView;
@property (nonatomic) UIView *screenBlockingButton;
@property (nonatomic) NSArray<NSLayoutConstraint *> *screenBlockingConstraints;
@property (nonatomic) NSString *screenBlockingSignature;

@end

#pragma mark -

@implementation ScreenLockViewController

- (void)loadView
{
    [super loadView];
    
    // Loki: Set gradient background
    self.view.backgroundColor = UIColor.clearColor;
    CAGradientLayer *layer = [CAGradientLayer new];
    layer.frame = UIScreen.mainScreen.bounds;
    UIColor *gradientStartColor = LKAppModeUtilities.isLightMode ? [UIColor colorWithRGBHex:0xF9F9F9] : [UIColor colorWithRGBHex:0x171717];
    UIColor *gradientEndColor = LKAppModeUtilities.isLightMode ? [UIColor colorWithRGBHex:0xFFFFFF] : [UIColor colorWithRGBHex:0x121212];
    layer.colors = @[ (id)gradientStartColor.CGColor, (id)gradientEndColor.CGColor ];
    [self.view.layer insertSublayer:layer atIndex:0];

    UIView *edgesView = [UIView containerView];
    [self.view addSubview:edgesView];
    [edgesView autoPinEdgeToSuperviewEdge:ALEdgeTop];
    [edgesView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
    [edgesView autoPinWidthToSuperview];

    UIImage *image = [UIImage imageNamed:@"SessionGreen64"];
    UIImageView *imageView = [UIImageView new];
    imageView.image = image;
    imageView.contentMode = UIViewContentModeScaleAspectFit;
    [edgesView addSubview:imageView];
    [imageView autoHCenterInSuperview];

    [imageView autoSetDimension:ALDimensionWidth toSize:64];
    [imageView autoSetDimension:ALDimensionHeight toSize:64];

    const CGFloat kButtonHeight = 40.f;
    OWSFlatButton *button =
        [OWSFlatButton buttonWithTitle:NSLocalizedString(@"Unlock Session", @"")
                                  font:[UIFont boldSystemFontOfSize:LKValues.mediumFontSize]
                            titleColor:LKAppModeUtilities.isLightMode ? UIColor.blackColor : UIColor.whiteColor
                       backgroundColor:UIColor.clearColor
                                target:self
                              selector:@selector(showUnlockUI)];
    [edgesView addSubview:button];

    [button autoSetDimension:ALDimensionHeight toSize:kButtonHeight];
    [button autoPinLeadingToSuperviewMarginWithInset:50.f];
    [button autoPinTrailingToSuperviewMarginWithInset:50.f];
    const CGFloat kVMargin = 65.f;
    [button autoPinBottomToSuperviewMarginWithInset:kVMargin];

    self.screenBlockingImageView = imageView;
    self.screenBlockingButton = button;

    [self updateUIWithState:ScreenLockUIStateScreenProtection isLogoAtTop:NO animated:NO];
}

// The "screen blocking" window has three possible states:
//
// * "Just a logo".  Used when app is launching and in app switcher.  Must match the "Launch Screen"
//    storyboard pixel-for-pixel.
// * "Screen Lock, local auth UI presented". Move the Signal logo so that it is visible.
// * "Screen Lock, local auth UI not presented". Move the Signal logo so that it is visible,
//    show "unlock" button.
- (void)updateUIWithState:(ScreenLockUIState)uiState isLogoAtTop:(BOOL)isLogoAtTop animated:(BOOL)animated
{
    OWSAssertIsOnMainThread();

    if (!self.isViewLoaded) {
        return;
    }

    BOOL shouldShowBlockWindow = uiState != ScreenLockUIStateNone;
    BOOL shouldHaveScreenLock = uiState == ScreenLockUIStateScreenLock;

    self.screenBlockingImageView.hidden = !shouldShowBlockWindow;

    NSString *signature = [NSString stringWithFormat:@"%d %d", shouldHaveScreenLock, isLogoAtTop];
    if ([NSObject isNullableObject:self.screenBlockingSignature equalTo:signature]) {
        // Skip redundant work to avoid interfering with ongoing animations.
        return;
    }

    [NSLayoutConstraint deactivateConstraints:self.screenBlockingConstraints];

    NSMutableArray<NSLayoutConstraint *> *screenBlockingConstraints = [NSMutableArray new];

    self.screenBlockingButton.hidden = !shouldHaveScreenLock;

    if (isLogoAtTop) {
        const CGFloat kVMargin = 60.f;
        [screenBlockingConstraints addObject:[self.screenBlockingImageView autoPinEdge:ALEdgeTop
                                                                                toEdge:ALEdgeTop
                                                                                ofView:self.view
                                                                            withOffset:kVMargin]];
    } else {
        [screenBlockingConstraints addObject:[self.screenBlockingImageView autoVCenterInSuperview]];
    }

    self.screenBlockingConstraints = screenBlockingConstraints;
    self.screenBlockingSignature = signature;

    if (animated) {
        [UIView animateWithDuration:0.35f
                         animations:^{
                             [self.view layoutIfNeeded];
                         }];
    } else {
        [self.view layoutIfNeeded];
    }
}

- (void)showUnlockUI
{
    OWSAssertIsOnMainThread();

    [self.delegate unlockButtonWasTapped];
}

- (BOOL)canBecomeFirstResponder
{
    return YES;
}

#pragma mark - Orientation

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskAllButUpsideDown;
}

@end

File: SignalUtilitiesKit/Screen Lock/ScreenLockViewController.swift
// Copyright © 2022 Rangeproof Pty Ltd. All rights reserved.

import UIKit
import SessionUIKit

open class ScreenLockViewController: UIViewController {
    public enum State {
        case none
        
        /// Shown while app is inactive or background, if enabled.
        case protection
        
        /// Shown while app is active, if enabled.
        case lock
    }
    
    public override var preferredStatusBarStyle: UIStatusBarStyle {
        return ThemeManager.currentTheme.statusBarStyle
    }
    
    public override var canBecomeFirstResponder: Bool { true }
    
    public var onUnlockPressed: (() -> ())?
    private var screenBlockingSignature: String?
    
    // MARK: - UI
    
    private let logoView: UIImageView = {
        let result: UIImageView = UIImageView(image: #imageLiteral(resourceName: "SessionGreen64"))
        result.contentMode = .scaleAspectFit
        result.isHidden = true
        
        return result
    }()
    
    public lazy var unlockButton: OutlineButton = {
        let result: OutlineButton = OutlineButton(style: .regular, size: .medium)
        result.translatesAutoresizingMaskIntoConstraints = false
        result.setTitle("Unlock Session", for: .normal)
        result.addTarget(self, action: #selector(showUnlockUI), for: .touchUpInside)
        result.isHidden = true
        
        // Need to match the launch screen so force the styling to be the primary green
        result.setThemeTitleColorForced(.primary(.green), for: .normal)
        result.setThemeBackgroundColorForced(.primary(.green), for: .highlighted)
        result.themeBorderColorForced = .primary(.green)

        return result
    }()
    
    // MARK: - Lifecycle
                                  
    public init(onUnlockPressed: (() -> ())? = nil) {
        self.onUnlockPressed = onUnlockPressed
        
        super.init(nibName: nil, bundle: nil)
    }
                                  
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
                                  
    open override func loadView() {
        super.loadView()
        
        view.themeBackgroundColor = .black  // Need to match the Launch screen

        let edgesView: UIView = UIView.container()
        self.view.addSubview(edgesView)
        edgesView.pin(to: view)
        
        edgesView.addSubview(logoView)
        logoView.center(in: edgesView)
        logoView.set(.width, to: 64)
        logoView.set(.height, to: 64)

        edgesView.addSubview(unlockButton)
        unlockButton.pin(.top, to: .bottom, of: logoView, withInset: Values.mediumSpacing)
        unlockButton.center(.horizontal, in: view)
        
        updateUI(state: .protection, animated: false)
    }
    
    /// The "screen blocking" window has three possible states:
    ///
    /// * "Just a logo".  Used when app is launching and in app switcher.  Must match the "Launch Screen" storyboard pixel-for-pixel.
    /// * "Screen Lock, local auth UI presented". Must match the "Launch Screen" storyboard pixel-for-pixel.
    /// * "Screen Lock, local auth UI not presented". Show "unlock" button.
    public func updateUI(state: State, animated: Bool) {
        guard isViewLoaded else { return }

        let shouldShowBlockWindow: Bool = (state != .none)
        let shouldHaveScreenLock: Bool = (state == .lock)

        self.logoView.isHidden = !shouldShowBlockWindow

        let signature: String = String(format: "%d", shouldHaveScreenLock)
        
        // Skip redundant work to avoid interfering with ongoing animations
        guard signature != self.screenBlockingSignature else { return }
        
        self.unlockButton.isHidden = !shouldHaveScreenLock
        self.screenBlockingSignature = signature

        guard animated else {
            self.view.setNeedsLayout()
            return
        }
        
        UIView.animate(withDuration: 0.3) { [weak self] in
            self?.view.layoutIfNeeded()
        }
    }
    
    @objc private func showUnlockUI() {
        self.onUnlockPressed?()
    }
}
